daemon/src/peer/run_state/running_waiting_room.rs

```diff
@@ -0,0 +1,345 @@
+        peer: PeerId,
+    },
+    /// Cloning failed for a urn and peer
+    CloningFailed {
+        /// The urn that failed
+        urn: Urn,
+        /// The peer we failed to clone from
+        peer: PeerId,
+        /// A description of why the cloning failed
+        reason: String,
+    },
+    /// Cloning succeeded for a urn and peer
+    Cloned {
+        /// The urn we cloned
+        urn: Urn,
+        /// The peer we cloned from
+        peer: PeerId,
+    },
+    /// A request for a urn was canceled
+    Canceled {
+        /// The urn that was canceled
+        urn: Urn,
+    },
+    /// A request was timed out
+    TimedOut {
+        /// The urn that timed out
+        urn: Urn,
+        /// The attempts that were made before the timeout
+        attempts: Option<usize>,
+    },
+    /// One tick of the waiting room
+    Tick,
+}
+
+/// `RunningWaitingRoom` is an adapter from the interface of `WaitingRoom` to
+/// the language of commands which is spoken by `RunState`. Whenever `RunState`
+/// needs to talk to `WaitingRoom` it does so via a wrapper method on
+/// `RunningWaitingRoom`. These wrapper methods contain the logic to convert
+/// the values  returned by `WaitingRoom` methods into `Vec<Command>`.
+pub(super) struct RunningWaitingRoom {
+    waiting_room: WaitingRoom<SystemTime, Duration>,
+}
+
+impl RunningWaitingRoom {
+    pub const fn new(waiting_room: WaitingRoom<SystemTime, Duration>) -> Self {
+        Self { waiting_room }
+    }
+
+    pub fn cancel(
+        &mut self,
+        urn: Urn,
+        timestamp: SystemTime,
+        sender: Sender<Result<Option<SomeRequest<SystemTime>>, WaitingRoomError>>,
+    ) -> Vec<Command> {
+        let state_before = self.waiting_room.requests();
+        match self.waiting_room.canceled(&urn, timestamp) {
+            Ok(()) => {
+                let request = self.waiting_room.remove(&urn);
+                let state_after = self.waiting_room.requests();
+                let transition = WaitingRoomTransition {
+                    timestamp,
+                    state_before,
+                    state_after,
+                    event: Event::Canceled { urn },
+                };
+                vec![
+                    Command::Control(command::Control::Respond(control::Response::CancelSearch(
+                        sender,
+                        Ok(request),
+                    ))),
+                    Command::PersistWaitingRoom(self.waiting_room.clone()),
+                    Command::EmitEvent(transition.into()),
+                ]
+            },
+            Err(e) => vec![
+                Command::Control(command::Control::Respond(control::Response::CancelSearch(
+                    sender,
+                    Err(e),
+                ))),
+                Command::PersistWaitingRoom(self.waiting_room.clone()),
+            ],
+        }
+    }
+
+    pub fn request(
+        &mut self,
+        urn: Urn,
+        time: SystemTime,
```

Comment:
nit: we have `timestamp`, `time`, and `now` namings for these parameters. Can we settle on one? :)

Reply:
Fixed